core/src/main/scala/kafka/log/TimeIndex.scala

```diff
@@ -179,18 +184,26 @@ class TimeIndex(_file: File, baseOffset: Long, maxIndexSize: Int = -1, writable:
     }
   }
 
+  override def resize(newSize: Int) {
+    inLock(lock) {
+      super.resize(newSize)
+      _lastEntry = lastEntryFromIndexFile
+    }
+  }
+
   /**
    * Truncates index to a known number of entries.
    */
   private def truncateToEntries(entries: Int) {
     inLock(lock) {
       _entries = entries
       mmap.position(_entries * entrySize)
+      _lastEntry = lastEntryFromIndexFile
     }
   }
 
   override def sanityCheck() {
-    val entry = lastEntry
+    val entry = lastEntryFromIndexFile
```

Comment:
Do we need to read from the file here? It seems we have already cached the result when constructing the `TimeIndex` and updated the cache if there is a change.

Reply:
Sure. I updated the code as suggested.